### PR TITLE
Monkeypatch for Qwen2-VL

### PR DESCRIPTION
**Pull request by tyler-romero**
## Summary
Monkeypatch for the recently-published [Qwen2-VL](https://huggingface.co/Qwen/Qwen2-VL-7B-Instruct).
HF `transformers` modeling code: https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen2_vl/modeling_qwen2_vl.py

Feature Request: https://github.com/linkedin/Liger-Kernel/issues/165

## Details
Qwen2-VL in `transformers` is available on `transformers` main but is yet to be published in a release.

## Testing Done
- Hardware Type: 4090
- [x] run `make test` to ensure correctness
- [x] run `make checkstyle` to ensure code style
- [x] run `make test-convergence` to ensure convergence

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -28,9 +28,9 @@
     keywords="triton,kernels,LLM training,deep learning,Hugging Face,PyTorch,GPU optimization",
     include_package_data=True,
     install_requires=[
-        "torch>=2.1.2",
-        "triton>=2.3.0",
-        "transformers>=4.42.0",
+        "torch>=2.4.0",
+        "triton>=3.0.0",
+        "transformers @ git+https://github.com/huggingface/transformers.git",
```
Comment (tyler-romero) [inline setup.py:33]: Will revert. Need to implement smarter importing of qwen2_vl dependencies.

File: src/liger_kernel/transformers/monkey_patch.py [hunk 2]
```diff
@@ -233,6 +234,54 @@ def apply_liger_kernel_to_qwen2(
...
+        lce_forward as qwen2_vl_lce_forward,
+    )
+
+    # Qwen2 VL has two rope implementations, neither of which is like liger_rotary_pos_emb
+    # if rope:
+    #     modeling_qwen2_vl.apply_multimodal_rotary_pos_emb = ...
+    #     modeling_qwen2_vl.apply_rotary_pos_emb_vision = ...
+    if rms_norm:
+        modeling_qwen2_vl.Qwen2RMSNorm = LigerRMSNorm
+    if layer_norm:
```
Comment (tyler-romero) [inline src/liger_kernel/transformers/monkey_patch.py:275]: First internal use of the LayerNorm kernel!

File: src/liger_kernel/transformers/monkey_patch.py [hunk 3]
```diff
@@ -233,6 +234,54 @@ def apply_liger_kernel_to_qwen2(
...
+        modeling_qwen2_vl.LayerNorm = LigerLayerNorm
+    if cross_entropy:
+        modeling_qwen2_vl.CrossEntropyLoss = LigerCrossEntropyLoss
+    if fused_linear_cross_entropy:
+        modeling_qwen2_vl.Qwen2VLForConditionalGeneration.forward = qwen2_vl_lce_forward
```
Comment (tyler-romero) [inline src/liger_kernel/transformers/monkey_patch.py:280]: Not AutoModelForCausalLMs like the others.

File: test/convergence/test_mini_models.py [hunk 4]
```diff
@@ -253,6 +258,50 @@
             attn_implementation="sdpa",  # default value, pytorch native attention
         ),
     ),
+    "mini_qwen2_vl": MiniModelConfig(
```
Comment (tyler-romero) [inline test/convergence/test_mini_models.py:261]: These tests test the llm portion of qwen2_vl, but not the vit portion (which LayerNorm affects)

**Comment by kostum123** [comment]
is this pr working? any show stoppers?

**Comment by tyler-romero** [comment]
It 'works' but not quite ready for merge yet b/c of lose bounds on convergence test, lack of testing for image inputs, and lack of shielding the imports for lower versions of `transformers`

File: test/convergence/test_mini_models.py [hunk 5]
```diff
@@ -29,8 +21,28 @@
...
+    DEFAULT_DATASET_PATH,
+    MiniModelConfig,
+    assert_verbose_allclose,
+    set_seed,
+    simple_collate_fn,
+    supports_bfloat16,
 )
 
+try:
+    # Qwen2-VL is only available in transformers>4.44.2
```
Comment (tyler-romero) [inline test/convergence/test_mini_models.py:36]: Tests pass w/ transformers @ HEAD on 4090

File: src/liger_kernel/transformers/monkey_patch.py [hunk 6]
```diff
@@ -233,6 +234,54 @@ def apply_liger_kernel_to_qwen2(
...
+    )
+
+    # TODO: Support Qwen2-VL's multimodal RoPE implementation
+
+    if rms_norm:
+        # https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen2_vl/modeling_qwen2_vl.py#L439
+        modeling_qwen2_vl.Qwen2RMSNorm = partial(
+            LigerRMSNorm, init_fn="ones", casting_mode="gemma"
+        )
+    if layer_norm:
```
Comment (tyler-romero) [inline src/liger_kernel/transformers/monkey_patch.py:275]: LayerNorm is only used in the ViT part of the model.

Q for reviewers: We don't currently have image inputs as part of the convergence test suite. Worth implementing? Should I just not support LayerNorm for now so we don't need to worry about it? Perhaps default to False?

Comment (ryankert01) [reply src/liger_kernel/transformers/monkey_patch.py:275]: I think the question is will we support more multi-modal? If yes, I think we should implement convergence test of it.

**Comment by tyler-romero** [comment]
All tests (including qwen2-vl) passing with:
```
    "torch>=2.4.0",
    "triton>=3.0.0",
    "transformers @ git+https://github.com/huggingface/transformers.git"
```

All tests (skipping qwen2-vl) passing without modifications to dependency versions.

**Comment by lancerts** [comment]
@ByronHsu

**Comment by ByronHsu** [comment]
I will take a look today and tomorrow

File: test/convergence/test_mini_models.py [hunk 7]
```diff
@@ -331,6 +331,7 @@ def run_mini_model(
 
     for i in range(num_steps):
         batch = next(loader_iter).to(model.device)
+        optimizer.zero_grad()
```
Comment (ByronHsu) [inline test/convergence/test_mini_models.py:334]: nit: should this be after line 337?

File: test/convergence/test_mini_models_no_logits.py [hunk 8]
```diff
@@ -305,6 +368,7 @@ def run_mini_model(
 
     for i in range(num_steps):
         batch = next(loader_iter).to(model.device)
+        optimizer.zero_grad()
```
Comment (ByronHsu) [inline test/convergence/test_mini_models_no_logits.py:371]: ditto

**Review by ByronHsu — APPROVED** [review]
lgtm

File: src/liger_kernel/transformers/__init__.py [hunk 9]
```diff
@@ -15,6 +15,7 @@
     apply_liger_kernel_to_mixtral,
     apply_liger_kernel_to_phi3,
     apply_liger_kernel_to_qwen2,
+    apply_liger_kernel_to_qwen2_vl,
```
Comment (ByronHsu) [inline src/liger_kernel/transformers/__init__.py:18]: will this break users if transformers < 4.44?

File: src/liger_kernel/transformers/model/qwen2_vl.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
+from typing import List, Optional, Tuple, Union
+
+import torch
+from torch.nn import CrossEntropyLoss
+from transformers.models.qwen2_vl.modeling_qwen2_vl import (
```
Comment (ByronHsu) [inline src/liger_kernel/transformers/model/qwen2_vl.py:5]: ditto. will it break users with lower version?

File: src/liger_kernel/transformers/__init__.py (hunk 9 above)
Comment (tyler-romero) [reply src/liger_kernel/transformers/__init__.py:18]: The import won't break, but if someone tries to call the function then they would get an import error as it tries to import the Qwen2-VL module internally.

File: src/liger_kernel/transformers/model/qwen2_vl.py (hunk 10 above)
Comment (tyler-romero) [reply src/liger_kernel/transformers/model/qwen2_vl.py:5]: Only if they explicitly import this module - and I would argue it should throw an ImportError in that case.

File: test/convergence/test_mini_models.py (hunk 7 above)
Comment (tyler-romero) [reply test/convergence/test_mini_models.py:334]: Anywhere in the loop that's not between backwards() and step() is valid - PyTorch tutorials tend to place it at the beginning of a training step.

**Review by lancerts — APPROVED** [review]

**Comment by ByronHsu** [comment]
Thanks for the effort!

**Pull request merged**